packages/ckb-indexer/tests/indexer.test.ts

```diff
@@ -37,7 +33,7 @@ test("subscribe cells", async (t) => {
     let spy = sinon.spy();
     const eventEmitter = indexer.subscribe(queryCase.queryOption);
     eventEmitter.on("changed", spy);
-    await new Promise((resulve) => setTimeout(resulve, 10000));
+    await new Promise((resulve) => setTimeout(resulve, 90000));
```

Comment:
why we should wait for the 90 seconds here?

Reply:
I have tried`while` and `Date.now()-startTime > 20000`.but  it does not work. `spy.callCount`  does not change after 2mins.
I think this is cause JS is Single Thread and `while` loop is always on exec stack,so there is no time for spy.callCount.

I planing to  use  `setTimeout` but  have any clear idea yet


-----------------------
there is the code i have tried

```javascript
test("subscribe cells", (t) => {
  let blockIndex = 0;
  const stub = sinon.stub(indexer, "tip").callsFake(() => {
    const blocks = JSON.parse(
      fs.readFileSync(path.join(__dirname, "./blocks_data.json")).toString()
    );
    const block = blocks[blockIndex];
    if (blockIndex !== 99) {
      blockIndex = blockIndex + 1;
    }
    return Promise.resolve({
      block_hash: block.header.hash,
      block_number: block.header.number,
    });
  });

  for (const queryCase of indexerSubscribeTestCases) {
    let spy = sinon.spy();
    const eventEmitter = indexer.subscribe(queryCase.queryOption);
    eventEmitter.on("changed", spy);
    const startTime = Date.now();
    let originSpyCount = spy.callCount;
    while (true) {
      if (originSpyCount !== spy.callCount) {
        console.log(spy.callCount);
        originSpyCount = spy.callCount;
      }
      if (
        Date.now() - startTime > 20000 ||
        spy.callCount === queryCase.expectedResult
      ) {
        console.log("spy.callCount", spy.callCount);
        t.is(spy.callCount, queryCase.expectedResult, queryCase.desc);
        stub.resetHistory();
        blockIndex = 0;
        spy.resetHistory();
        break;
      }
    }
  }
});

```

Reply:
I think you need a retry helper like this

```ts
async function asyncRetry(
  callback: () => Promise<boolean> | boolean,
  interval: number,
  timeout: number
) {
  const start = Date.now();
  while (true) {
    if (Date.now() - start >= timeout) throw new Error("timeout");
    const shouldBreak = await callback();
    if (shouldBreak) break;
    await new Promise((resolve) => setTimeout(resolve, interval));
  }
}
```

usage demo

```ts
function main() {
  let callCount = 0;

  asyncRetry(
    () => {
      if (callCount >= 3) return true;
      console.log("count:", callCount);
      callCount += 1;
      return false;
    },
    1000,
    10000
  ).then(() => {
    console.log(callCount);
  });
}

main();

```